src/Affix.js

```diff
@@ -0,0 +1,185 @@
+import classNames from 'classnames';
+import ownerDocument from 'dom-helpers/ownerDocument';
+import getHeight from 'dom-helpers/query/height';
+import getOffset from 'dom-helpers/query/offset';
+import getOffsetParent from 'dom-helpers/query/offsetParent';
+import getScrollTop from 'dom-helpers/query/scrollTop';
+import requestAnimationFrame from 'dom-helpers/util/requestAnimationFrame';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import addEventListener from './utils/addEventListener';
+import getDocumentHeight from './utils/getDocumentHeight';
+import ownerWindow from './utils/ownerWindow';
+
+class Affix extends React.Component {
+  constructor(props, context) {
+    super(props, context);
+
+    this.state = {
+      affixed: 'top',
+      affixStyle: null
+    };
+
+    this._needPositionUpdate = false;
+  }
+
+  componentDidMount() {
+    const node = ReactDOM.findDOMNode(this);
+    this._isMounted = true;
+
+    this._windowScrollListener = addEventListener(
+      window, 'scroll', () => this.onWindowScroll()
+    );
+    this._documentClickListener = addEventListener(
+      ownerDocument(node), 'click', () => this.onDocumentClick()
+    );
+
+    this.positionTopBase = getOffset(node).top;
+    this.onPositionUpdate();
+  }
+
+  componentWillReceiveProps() {
+    this._needPositionUpdate = true;
+  }
+
+  componentDidUpdate() {
+    if (this._needPositionUpdate) {
+      this._needPositionUpdate = false;
+      this.onPositionUpdate();
+    }
+  }
+
+  componentWillUnmount() {
+    if (this._windowScrollListener) {
+      this._windowScrollListener.remove();
+    }
+    if (this._documentClickListener) {
+      this._documentClickListener.remove();
+    }
+
+    this._isMounted = false;
+  }
+
+  onWindowScroll() {
+    this.onPositionUpdate();
+  }
+
+  onDocumentClick() {
+    requestAnimationFrame(() => this.onPositionUpdate());
```

Comment:
recalculate on document click? is that just sort of a shotgun mutation observer :P or is there a more subtle thing I'm missing here?